nodestats.go

```diff
@@ -0,0 +1,106 @@
+package mir
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/mir/pkg/logging"
+	t "github.com/filecoin-project/mir/pkg/types"
+	"github.com/filecoin-project/mir/pkg/util/maputil"
+)
+
+// ==============================================================================================================
+// Event dispatching statistics
+// ==============================================================================================================
+
+// eventDispatchStats saves statistical information about the dispatching of events between modules,
+// such the numbers of events dispatched for each module.
+type eventDispatchStats struct {
+	dispatchCounts map[t.ModuleID]int
+	eventCounts    map[t.ModuleID]int
+	numDispatches  int
+	lastDispatch   time.Time
+}
+
+// newDispatchStats returns a new eventDispatchStats object with all counters set to 0
+// and the last dispatch time set to the current time.
+func newDispatchStats(moduleIDs []t.ModuleID) eventDispatchStats {
+	stats := eventDispatchStats{
+		dispatchCounts: make(map[t.ModuleID]int),
+		eventCounts:    make(map[t.ModuleID]int),
+		numDispatches:  0,
+		lastDispatch:   time.Now(),
+	}
+	for _, moduleID := range moduleIDs {
+		stats.dispatchCounts[moduleID] = 0
+		stats.eventCounts[moduleID] = 0
+	}
+	return stats
+}
+
+// AddDispatch registers the dispatching of an event list between two modules to the statistics.
+func (ds *eventDispatchStats) AddDispatch(mID t.ModuleID, numEvents int) {
+	ds.numDispatches++
+	ds.dispatchCounts[mID]++
+	ds.eventCounts[mID] += numEvents
+}
+
+// CombinedStats returns the dispatching statistics combined with information about event buffer occupancy
+// in a format that can directly be passed to the logger.
+// For each module, it includes the number of event lists dispatched to that module (d),
+// the total number of events in those lists (e),
+// and the number of events still buffered at the input of that module (b).
+func (ds *eventDispatchStats) CombinedStats(bufferStats map[t.ModuleID]int) []interface{} {
+	logVals := make([]interface{}, 0, len(ds.eventCounts)+2)
+	totalEventsDispatched := 0
+	totalEventsBuffered := 0
+	for mID, cnt := range ds.dispatchCounts {
+		logVals = append(logVals,
+			fmt.Sprint(mID), fmt.Sprintf("d(%d)-e(%d)-b(%d)", cnt, ds.eventCounts[mID], bufferStats[mID]),
```

Comment:
missing `\n`?

Reply:
No, this is deliberate too. This string will be used as an argument to the `Log` function and should not end with a new line.